staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/testing/clock/event_clock.go

```diff
@@ -0,0 +1,221 @@
+	go func() {
+		select {
+		case t := <-ch:
+			f(t)
+		}
+	}()
+}
+
+// EventAfterTime schedules an EventFunc
+func (r RealEventClock) EventAfterTime(f EventFunc, t time.Time) {
+	now := time.Now()
+	d := t.Sub(now)
+	if d <= 0 {
+		go f(now)
+	} else {
+		r.EventAfterDuration(f, d)
+	}
+}
+
+// waitGroupCounter is a wait group used for a GoRoutine Counter.  This private
+// type is used to disallow direct waitGroup access
+type waitGroupCounter struct{ sync.WaitGroup }
+
+// FakeEventClock is one whose time does not pass implicitly but
+// rather is explicitly set by invocations of its SetTime method
+type FakeEventClock struct {
+	clock.FakePassiveClock
+
+	// waiters is a heap of waiting work, sorted by time
+	waiters     eventWaiterHeap
+	waitersLock sync.RWMutex
+
+	// clientWG may be nil and if not supplies constraints on time
+	// passing in Run.  The Run method will not pick a new time until
+	// this is nil or its counter is zero.
+	clientWG *waitGroupCounter
+
+	// fuzz is the amount of noise to add to scheduling.  An event
+	// requested to run at time T will run at some time chosen
+	// uniformly at random from the interval [T, T+fuzz]; the upper
+	// bound is exclusive iff fuzz is non-zero.
+	fuzz time.Duration
+
+	// rand is the random number generator to use in fuzzing
+	rand *rand.Rand
+}
+
+type eventWaiterHeap []eventWaiter
+
+var _ heap.Interface = (*eventWaiterHeap)(nil)
+
+type eventWaiter struct {
+	targetTime time.Time
+	f          EventFunc
+}
+
+// NewFakeEventClock constructor.  The given `r *rand.Rand` must
+// henceforth not be used for any other purpose.  If `r` is nil then a
+// fresh one will be constructed, seeded with the current real time.
+// The clientWG can be `nil` and if not is used to let Run know about
+// additional work that has to complete before time can advance.
+func NewFakeEventClock(t time.Time, fuzz time.Duration, r *rand.Rand) (*FakeEventClock, counter.GoRoutineCounter) {
+	grc := &waitGroupCounter{}
+
+	if r == nil {
+		r = rand.New(rand.NewSource(time.Now().UnixNano()))
+		r.Uint64()
+		r.Uint64()
+		r.Uint64()
+	}
+	return &FakeEventClock{
+		FakePassiveClock: *clock.NewFakePassiveClock(t),
+		clientWG:         grc,
+		fuzz:             fuzz,
+		rand:             r,
+	}, grc
+}
+
+// GetNextTime returns the next time at which there is work scheduled,
+// and a bool indicating whether there is any such time
+func (fec *FakeEventClock) GetNextTime() (time.Time, bool) {
+	fec.waitersLock.RLock()
+	defer fec.waitersLock.RUnlock()
+	if len(fec.waiters) > 0 {
+		return fec.waiters[0].targetTime, true
+	}
+	return time.Time{}, false
+}
+
+// Run runs all the events scheduled, and all the events they
+// schedule, and so on, until there are none scheduled or the limit is not
+// nil and the next time would exceed the limit.  The clientWG given in
+// the constructor gates each advance of time.
+func (fec *FakeEventClock) Run(limit *time.Time) {
+	for {
+		if fec.clientWG != nil {
```

Comment:
This test is not needed, the condition is always true.

Reply:
removed the nil check